apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -1093,11 +1104,80 @@ call_query(QM, Id, Index, Ref, Query, QueryOpts) ->
         {ok, _Group, #{status := ?status_connecting, error := unhealthy_target}} ->
             {error, {unrecoverable_error, unhealthy_target}};
         {ok, _Group, Resource} ->
-            do_call_query(QM, Id, Index, Ref, Query, QueryOpts, Resource);
+            set_rule_id_trace_meta_data(Query),
+            QueryResult = do_call_query(QM, Id, Index, Ref, Query, QueryOpts, Resource),
+            %% do_call_query does not throw an exception as the call to the
+            %% resource is wrapped in a try catch expression so we will always
+            %% unset the trace meta data
+            unset_rule_id_trace_meta_data(),
+            QueryResult;
         {error, not_found} ->
             ?RESOURCE_ERROR(not_found, "resource not found")
     end.
 
+set_rule_id_trace_meta_data(Requests) when is_list(Requests) ->
+    %% Get the rule ids from requests
+    RuleIDs = lists:foldl(fun collect_rule_id/2, #{}, Requests),
+    ClientIDs = lists:foldl(fun collect_client_id/2, #{}, Requests),
+    StopAfterRender = lists:foldl(fun collect_stop_after_render/2, no_info, Requests),
+    StopAfterRenderVal =
+        case StopAfterRender of
+            only_true ->
+                logger:update_process_metadata(#{stop_action_after_render => false}),
+                true;
+            only_false ->
+                false;
+            mixed ->
```

Comment:
can modify do_collect_requests to avoid collecting mixed requests in one batch

Reply:
> can modify do_collect_requests to avoid collecting mixed requests in one batch

Sure, this is possible but is it something that we want to do? This might introduce complex logic and some overhead for something that will be very rare in practice (at least if we document that this problem can arise when setting stop_action_after_render).

If we want it, does it need to be implemented before we merge this PR or can it be introduced later?

Reply:
I’m sure it will be less complex than this collect logic. 
the batch collection is pretty straightforward.

Reply:
Better to fix it before merge it

Reply:
Ok, I will look into it.

Reply:
Thanks for the discussion. Fixed here [cf56050](https://github.com/emqx/emqx/pull/12827/commits/cf56050759d96806b1e1cb39a564cdf7f3acbb44)